packages/peregrine/lib/talons/CheckoutPage/useShippingMethod.js

```diff
@@ -159,9 +167,67 @@ export const useShippingMethod = props => {
         // Determine the component's display state.
         const nextDisplayState = selectedMethod
             ? displayStates.DONE
+            : loading || isBackgroundAutoSelecting
+            ? displayStates.INITIALIZING
             : displayStates.EDITING;
         setDisplayState(nextDisplayState);
-    }, [data]);
+    }, [data, isBackgroundAutoSelecting, loading]);
+
+    // If an authenticated user does not have a preferred shipping method,
+    // auto-select the least expensive one for them.
+    useEffect(() => {
+        if (!data) return;
+        if (!cartId) return;
+        if (!isSignedIn) return;
+
+        // Functions passed to useEffect should be synchronous.
+        // Set this helper function up as async so we can wait on the mutation
+        // before re-querying.
+        const autoSelectShippingMethod = async shippingMethod => {
+            const { carrier_code, method_code } = shippingMethod;
+
+            setIsBackgroundAutoSelecting(true);
+
+            // Perform the operation on the backend.
+            await setShippingMethodCall({
```

Comment:
Tommy is right - if the data can be returned by a mutation and assuming the response data is correct (as in no bug in GQL) we should opt to query for that data in the response rather than firing a second request/query. If you're able to get all you need in the response you can get rid of the extra loading state variable too. Seems like a win-win!